File: advent_cpp/day1/driver.cpp
#include <fstream>
#include <string>
#include <vector>

#include "puzzle2.cpp"

int main(int argc, char *argv[])
{
    const int NUMBER_REPEATS = 5;
    const int NUMBER_TRIALS = 500;
    const int P1_SOLUTION = 1882714;
    const int P2_SOLUTION = 19437052;
    const std::string file_name = argv[1];

    std::vector<long> p1_exec_times;
    for (int i = 0; i < NUMBER_REPEATS; i++)
    {
        auto start = std::chrono::high_resolution_clock::now();
        for (int j = 0; j < NUMBER_TRIALS; j++) 
        {
            puzzle1Solver(file_name);
        }
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        p1_exec_times.push_back(duration.count() / NUMBER_TRIALS);
    }
    if (puzzle1Solver(file_name) == P1_SOLUTION)
    {
        std::cout << "Puzzle 1: " << *std::min_element(p1_exec_times.begin(), p1_exec_times.end()) << " microseconds" << std::endl;
    }
    else
    {
        std::cout << "Puzzle 1: incorrect" << std::endl;
    }

    std::vector<long> p2_exec_times;
    for (int i = 0; i < NUMBER_REPEATS; i++)
    {
        auto start = std::chrono::high_resolution_clock::now();
        for (int j = 0; j < NUMBER_TRIALS; j++) 
        {
            puzzle2Solver(file_name);
        }
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        p2_exec_times.push_back(duration.count() / NUMBER_TRIALS);
    }
    int p2_solution = puzzle2Solver(file_name);
    if (p2_solution == P2_SOLUTION)
    {
        std::cout << "Puzzle 2: " << *std::min_element(p2_exec_times.begin(), p2_exec_times.end()) << " microseconds" << std::endl;
    }
    else
    {
        std::cout << "Puzzle 2: incorrect" << std::endl;
    }

    std::vector<long> p2s_exec_times;
    for (int i = 0; i < NUMBER_REPEATS; i++)
    {
        auto start = std::chrono::high_resolution_clock::now();
        for (int j = 0; j < NUMBER_TRIALS; j++) 
        {
            puzzle2SortedSolver(file_name);
        }
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        p2s_exec_times.push_back(duration.count() / NUMBER_TRIALS);
    }
    int p2s_solution = puzzle2SortedSolver(file_name);
    if (p2s_solution == P2_SOLUTION)
    {
        std::cout << "P2sorted2: " << *std::min_element(p2s_exec_times.begin(), p2s_exec_times.end()) << " microseconds" << std::endl;
    }
    else
    {
        std::cout << "P2sorted: incorrect" << std::endl;
    }
}

File: advent_python/day1/driver.py
import timeit
import sys

from puzzle1 import puzzle1_solver
from puzzle2 import puzzle2_solver, puzzle2_memod_solver

def main(file_name):
    NUMBER_REPEATS = 5
    NUMBER_TRIALS = 10000
    TIME_ROUND = 2
    P1_SOLUTION = 1882714
    P2_SOLUTION = 19437052
    
    p1_time = min(timeit.repeat(lambda: puzzle1_solver(file_name), number=NUMBER_TRIALS, repeat=NUMBER_REPEATS))
    p1_solution = puzzle1_solver(file_name)

    p2_time = min(timeit.repeat(lambda: puzzle2_solver(file_name), number=NUMBER_TRIALS, repeat=NUMBER_REPEATS))
    p2_solution = puzzle2_solver(file_name)

    p2_time_memoized = min(timeit.repeat(lambda: puzzle2_memod_solver(file_name), number=NUMBER_TRIALS, repeat=NUMBER_REPEATS))
    p2_solution_memoized = puzzle2_memod_solver(file_name)

    if p1_solution == P1_SOLUTION:
        print("Puzzle 1:", round(p1_time / NUMBER_TRIALS * 1000000, TIME_ROUND), "microseconds")
    else:
        print("Puzzle 1: incorrect")
    if p2_solution == P2_SOLUTION:
        print("Puzzle 2:", round(p2_time / NUMBER_TRIALS * 1000000, TIME_ROUND), "microseconds")
    else:
        print("Puzzle 2: incorrect")
    if p2_solution_memoized == P2_SOLUTION:
        print("P2 memod:", round(p2_time_memoized / NUMBER_TRIALS * 1000000, TIME_ROUND), "microseconds")
    else:
        print("P2 memod: incorrect")

if __name__ == '__main__':
    main(sys.argv[1])